src/utils.ts

```diff
@@ -509,111 +513,1346 @@ function decimalizationInternal(value: string, systemNumeration: number) {
-    originalFunctionScope.varArray.forEach((v, index) => {
-        if (v.varName == '@context') {
-            const context = newFuncScope.findVariable('@context') as Variable;
-            context.setVarIndex(v.varIndex);
-            context.scope = newFuncScope;
-        } else {
-            const varType = v.varType;
-            const new_var = new Variable(
-                v.varName,
-                varType,
-                [],
-                v.varIndex,
-                v.isLocalVar(),
-                v.initExpression,
+    if (
+        !specializedMethodType.genericOwner ||
+        (specializedMethodType.genericOwner &&
+            !specializedMethodType.genericOwner.belongedScope) ||
+        !typeArguments
+    )
+        return;
+
+    const genericMethodType = specializedMethodType.genericOwner as TSFunction;
+    const genericMethodScope =
+        genericMethodType.belongedScope! as FunctionScope;
+    const newMethodScope = new FunctionScope(parentScope);
+    genericMethodScope.specialize(newMethodScope);
+    newMethodScope.setClassName(classType.className);
+    newMethodScope.setFuncName(genericMethodScope.getName());
+
+    newMethodScope.setGenericOwner(genericMethodScope);
+    genericMethodScope.addSpecializedScope(classType.className, newMethodScope);
+
+    newMethodScope.setFuncType(specializedMethodType);
+    specializedMethodType.setBelongedScope(newMethodScope);
+
+    const typeParameters = genericMethodType.belongedClass!.typeArguments!;
+    if (!typeArguments) return;
+
+    // process funcName, mangledName and className of FunctionScope
+    if (genericMethodScope.mangledName !== '') {
+        const genericMangledName = genericMethodScope.mangledName;
+        const reverse = genericMangledName.split('|').reverse();
+        // class name
+        reverse[1] = classType.className;
+        newMethodScope.mangledName = reverse.reverse().join('|');
+    }
+
+    // process function parameters and create scope
+    for (let idx = 0; idx < genericMethodType.getParamTypes().length; idx++) {
+        const genericParamType = genericMethodType.getParamTypes()[idx];
+        if (
+            genericParamType instanceof TSTypeWithArguments &&
+            genericParamType.belongedScope
+        )
+            createScopeBySpecializedType(
+                specializedMethodType.getParamTypes()[
+                    idx
+                ] as TSTypeWithArguments,
+                genericParamType.belongedScope.parent!,
+                context,
             );
-            new_var.scope = newFuncScope;
-            newFuncScope.addVariable(new_var);
+    }
+    return newMethodScope;
+}
+
+/**
+ * @describe create a new specialize FunctionScope
+ * @param specializedFunctionType the new specialized function type
+ * @param parentScope the parent scope
+ * @param context the parser context
+ * @returns a new specialized FunctionScope
+ */
+function createFunctionScope(
```

Comment:
There is not big difference between `createFunctionScope` and `createMethodScope`, can we use a common function to represent?

Reply:
It is possible to merge them together. `createMethodScope` will have some logic for processing **_Class_**. I encapsulated them separately, just to distinguish **method** and **function**.

Reply:
Yeah, and how about merge them, and use a parameter to distinguish them? since their inner logic is very close.

Reply:
done